Pass on the Layout chapter (#12)
pavpanchekha/emberfox: book/styles.md

sort the rules in priority order, with higher-priority rules first.

So let's add a `priority` method to the selector classes that return
this priority. In this simplest implementation the exact numbers don't
matter if they sort right, but with an eye toward the future let's
assign tag selectors priority `1`, class selectors priority `16`, and
id selectors priority `256`:

``` {.python}
class TagSelector:
    def priority(self):
        return 1
        
class ClassSelector:
    def priority(self):
        return 16
        
class IdSelector:
    def priority(self):
        return 256
```

Now, before you call `style`, you should sort your list of rules:

``` {.python}
rules.sort(key=lambda (selector, body): selector.priority(), reverse=True)
```

Note the `reverse` flag: we want higher-priority rules to come first.
In Python, the `sort` function is *stable*, which means that things
keep their relative order if possible. This means that in general, a
later rule has higher priority, unless the selectors used force
something different.

Downloading styles
==================

Browsers get CSS code from two sources. First, each browser ships with a

> *browser style sheet*, which defines the default styles for all sorts of

Reviewer: ```suggestion
*user agent style sheet*, which defines the default styles for all sorts of
```

Reviewer: Did you say yet what a "user agent" is?

take priority over the browser style sheet.^[In reality this is
handled by browser styles having a lower score than user styles in the
cascade order, but our browser style sheet only has tag selectors in
it, so every rule already has the lowest possible score.] With the
rules loaded, we need only sort and apply them and then do layout:

``` {.python}
def load(self, url):
    # ...
    rules.sort(key=lambda (selector, body): selector.priority(),
        reverse=True)
    style(nodes, rules)
    self.layout(nodes)
```

With this done, each page should now automatically apply the margins
and paddings specified in the browser stylesheet, making it possible
to delete some of `InlineLayout`'s tag handlers in its `close` method.

Inherited styles
================

Our implementation of margins, borders, and padding styles only affect
the block layout mode.[^inline-margins] We'd like to extend CSS to
affect inline layout mode as well, for example to change text styling.
But there's a catch: inline layout is mostly concerned with text, but
text nodes don't have any styles at all. How can that work?

[^inline-margins]: Margins, borders, and padding can be applied to

>     inline layout objects in a real browser, but they work kind of a

Reviewer: ```suggestion
    inline layout objects in a real browser, but they work in a kind of
```

        # ...
```

With `font-weight` and `font-style` set on every node, `InlineLayout`
no longer needs `style`, `weight`, and `size` fields; they were only
there to track when text was inside or outside `<i>` and `<b>` tags,
and now styles and inheritance are doing that job:

``` {.python}
class InlineLayout:
    def font(self):
        bold = node.style["font-weight"]
        italic = node.style["font-style"]
        if italic == "normal": italic = "roman"
        size = int(px(node.style.get("font-size")) * .75)
        return tkinter.font.Font(size=size, weight=weight, slant=slant)
    
    def text(self, text):
        font = self.font()
        for word in text.split():
            # ...
```

Note that the `font-style` needs to replace the CSS default of
"normal" with the Tk value "roman", and the `font-size` needs to be
converted from points to pixels.[^72ppi]

[^72ppi]: Normally you think of points as a physical length unit (one
    72^nd^ of an inch) and pixels as a digital unit (dependent on the
    screen) but in CSS, the conversion is fixed at exactly 75% (or 96

>     pixels per inch). I'm not sure why, it seems weird and it does

Reviewer: The reason is to ensure existing sites are readable on all devices.

Reviewer: Historically, it seems to have gone like this:

1. most/all monitors had similar pixel densities
2. sites were build-and-forget
3. CSS introduced the 'zoom' property, which allowed scaling outside of the monitor, and laid bare the non-physicality of "physical" units
4. High-DPI/"retina" displays futher moved way from true "phyisical"

    cause problems.

Now support for the `i`, `b`, `small`, and `big` tags can all be moved
to CSS:

``` {.css}
i { font-style: italic; }
b { font-weight: bold; }
small { font-size: 12px; }
big { font-size: 20px; }
```

Another place where the code depends on specific tag names is
`has_block_children`, which relies on a list of inline elements. The
CSS `display`, which can be either `block` or `inline`, replaces that
mechanism.^[Modern CSS adds way more values, like `run-in` or
`inline-block` or `flex` or `grid`, and it has layout modes set by

> other properties, like `float` and `position`. Design matters.]

Reviewer: Not sure what you are trying to get across by saying "design matters".

So we can add all the inline elements to our browser style sheet:

``` {.css}
a { display: inline; }
em { display: inline; }
/* ... */
```

And then read that in `has_block_children`:

``` {.python}
def has_block_children(self):
    for child in self.node.children:
        # ...
        elif child.style.get("display", "inline") == "inline":
            return False
    return True
```

With these changes, `InlineLayout` can lose its `open` and `close`
methods, becoming a small, self-contained engine for line layout while
most of its domain-specific knowledge of tags is moved to the browser
style sheet.

That style sheet is easier to edit, since it's independent of the rest
of the code. And while sometimes moving things to a data file means

> maintaining a new format. Here we get to reuse a format, CSS, that our

Reviewer: ```suggestion
maintaining a new format, here we get to reuse a format, CSS, that our
```

browser needs to support anyway.

Summary
=======

This chapter implemented a rudimentary but complete styling engine,
including downloading, parser, matching, sorting, and applying CSS
files. That means we:

- Added styling support in both `style` attributes and `link`ed CSS files;
- Implemented for margins, borders, and padding to block layout objects;
- Refactored `InlineLayout` to move the font properties to CSS;
- Removed most tag-specific reasoning from our layout code.

Our styling engine is also relatively easy to extend with properties
and selectors.

Exercises
=========

> *Shortcuts*: CSS "shortcut properties" set multiple related CSS

Reviewer: s/shortchut/shorthand/

properties at the same time; `margin`, `padding`, `border-width`, and

> `font` are all popular shortcuts. Implement these four shortcut

Reviewer: Add examples, like "margin: 3px 4px 5px 6px" means top right bottom left.

properties. If you do it in the `body` parsing function, you won't
need to change the rest of the code. Start with the case where all of

> the subproperties are specified, then add default values.

Reviewer: by "default values", do you mean supporting the "auto" keyword?

> *Comma*: CSS allows a rule to have multiple, comma-separated

Reviewer: Is there an official terminology for this we can use instead?

selectors, which is basically the same as multiple rules with the same

> body. Implement this and use it to shorten and shorten the browser

Reviewer: Give an example.

style sheet.

*Width/Height*: Add support to block layout objects for the `width`
and `height` properties. These can either be a pixel value, which
directly sets the width or height of the layout object, or the word
`auto`, in which case the existing layout algorithm is used.

*Percentages*: Most places where you can specify a pixel value in CSS,
you can also write a percentage value like `50%`. When you do that for
`margin`, `border`, or `padding` properties, it's relative to the
layout object's width, while when you do it for `font-weight` it's
relative to the parent's font size. Implement percentage values for
all of these properties.

> *Combinations*: Sometimes you want to select an element by tag *and*

Reviewer: Same question here - better to use official terminology.

> class. You do this by concatenating the selectors without anything in

Reviewer: ```suggestion
class. You do this by concatenating the selectors without any characters (including whitespace) in
```

between: `span.announce` selects elements that match both `span` and

> `.announce`. Implement a new `AndSelector` class to represent these

Reviewer: Add an example.

and modify the parser to parse them. Sum priorities.[^lexicographic]

[^lexicographic]: You're supposed to use lexicographic scoring for
    these `AndSelector` things, but sums will work fine as long as no
    one strings more than 16 selectors together.

> *Descendants*: When multiple selectors are separated with spaces, like

Reviewer: ```suggestion
*Descendant selectors*: When multiple selectors are separated with spaces, like
```

`ul b`, that selects all `<b>` elements with a `<ul>` ancestor.
Implement descendent selectors; scoring for descendent selectors works
just like for comnbination selectors. Make sure that something like `section
.warning` selects warnings inside sections, while `section.warning`

> selects warnings that *are* sections.

Reviewer: Another fun exercise:

*Ancestor selectors*: There is no way in CSS at present to do the inverse of descendant selectors. For example, there is no way to have a CSS rule that matches "all ul tags that contain a b in the subtree". Can you guess why? Try to implement ancestor selectors and analyze the speed of your algorithm.
